src/python/pycold/_param_validation.py

```diff
@@ -0,0 +1,653 @@
+                    f"{', '.join([str(c) for c in constraints[:-1]])} or"
+                    f" {constraints[-1]}"
+                )
+
+            raise ValueError(
+                f"The {param_name!r} parameter of {caller_name} must be"
+                f" {constraints_str}. Got {param_val!r} instead."
+            )
+
+
+def make_constraint(constraint):
+    """Convert the constraint into the appropriate Constraint object.
+    Parameters
+    ----------
+    constraint : object
+        The constraint to convert.
+    Returns
+    -------
+    constraint : instance of _Constraint
+        The converted constraint.
+    """
+    if isinstance(constraint, str) and constraint == "array-like":
+        return _ArrayLikes()
+    if isinstance(constraint, str) and constraint == "random_state":
+        return _RandomStates()
+    if constraint is callable:
+        return _Callables()
+    if constraint is None:
+        return _NoneConstraint()
+    if isinstance(constraint, type):
+        return _InstancesOf(constraint)
+    if isinstance(constraint, (Interval, StrOptions, Options, HasMethods)):
+        return constraint
+    if isinstance(constraint, str) and constraint == "boolean":
+        return _Booleans()
+    if isinstance(constraint, str) and constraint == "verbose":
+        return _VerboseHelper()
+    if isinstance(constraint, str) and constraint == "missing_values":
+        return _MissingValues()
+    raise ValueError(f"Unknown constraint type: {constraint}")
+
+
+def validate_params(parameter_constraints):
+    """Decorator to validate types and values of functions and methods.
+    Parameters
+    ----------
+    parameter_constraints : dict
+        A dictionary `param_name: list of constraints`. See the docstring of
+        `validate_parameter_constraints` for a description of the accepted constraints.
+        Note that the *args and **kwargs parameters are not validated and must not be
+        present in the parameter_constraints dictionary.
+    Returns
+    -------
+    decorated_function : function or method
+        The decorated function.
+    """
+
+    def decorator(func):
+        # The dict of parameter constraints is set as an attribute of the function
+        # to make it possible to dynamically introspect the constraints for
+        # automatic testing.
+        setattr(func, "_skl_parameter_constraints", parameter_constraints)
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+
+            func_sig = signature(func)
+
+            # Map *args/**kwargs to the function signature
+            params = func_sig.bind(*args, **kwargs)
+            params.apply_defaults()
+
+            # ignore self/cls and positional/keyword markers
+            to_ignore = [
+                p.name
+                for p in func_sig.parameters.values()
+                if p.kind in (p.VAR_POSITIONAL, p.VAR_KEYWORD)
+            ]
+            to_ignore += ["self", "cls"]
```

Comment:
In general the first params of a method or classmethod don't have to be `self` or `cls`, so it's very easy to break this check.

The better way to do this would be to have a variant of this decorator for functions and another one for a method. (like [ubelt.memoize and ubelt.memoize_method](https://ubelt.readthedocs.io/en/latest/ubelt.util_memoize.html)). If this is only meant for methods, then simply say so and ignore the zero-th argument, which is a better condition (i.e. it is a method and the zeroth arg) for the `self` or `cls` variable.

Another option is do something like I do in [ubelt.compatible](https://ubelt.readthedocs.io/en/latest/ubelt.util_func.html#ubelt.util_func.compatible) where the user can specify a `start` param which is set to 1 if it's a method and 0 if its a function.

Reply:
Ignore my comments here. It's scikit-learn code, so probably best to leave this as-is.